consensus/engine.go

```diff
@@ -86,6 +88,15 @@ const (
 	On  = true
 )
 
+// REVIEW:
+//  Are we sure we want to make the Start blocking for the whole lifecycle of the application?
+//  If so, it's also not clear why we would need a seperate Stop method, if the caller can just cancel
+//  the context given to Start to stop the Engine.
+//
+//  This is also like the central application loop, which I would move to a higher level object.
+//  I think we should follow single responsibility here any move this loop to an entity responsible
+//  for this operation.
```

Comment:
I think the `Engine` is the right object/level to perform this loop. Engine implies something that you start and is continually running rather than perhaps a single consensus `Instance` which runs until the termination criteria of the algorithm are met.

As to `Start`/`Stop` lifecycle management, I agree that I kind of have two designs pasted together:
- The non-blocking `Start` and `Stop` methods
- The `Run(ctx)` method that uses the context to start and stop it. 

I guess my deciding to somewhat support both is that it gives flexibility for how the user may use it.

I think when we move on to writing out the ABCI socket layer we will have a better idea of how the `Engine` should be run.